### PR TITLE
[FLINK-23478][k8s] Guarantee the watching event handling order

### PR DESCRIPTION
**Pull request by yittg**
## What is the purpose of the change

Guarantee the watching event handling order manually in KubernetesSharedInformer instead of depending on a lock.

## Verifying this change

This change is already covered by existing tests, such as `KubernetesSharedInformerITCase`.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

**Comment by yittg** [comment]
cc @wangyang0918 @dmvk @tillrohrmann

File: flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/resources/KubernetesSharedInformer.java
Comment (dmvk) [inline flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/resources/KubernetesSharedInformer.java:281]: This makes sense 👍 Have you been able to verify locally that this fixes the issue?

I'm just thinking whether executing callbacks in a loop is ok (does it have potential to block the thread for a long time?), or we should execute callbacks one by one (each callback gets a new Runnable). IMO it's ok as is, just a first thought I had when I was looking at the code.

Comment (dmvk) [reply flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/resources/KubernetesSharedInformer.java:281]: nit: We may want to use a blocking poll here.

Comment (yittg) [reply flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/resources/KubernetesSharedInformer.java:281]: no need blocking poll, the poll can always get a element.

Comment (yittg) [reply flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/resources/KubernetesSharedInformer.java:281]: I've verify locally for many rounds, no exception ever found so far.

The execution thread will only survive for a short period only if new event comes before the last event complete.

Comment (yittg) [reply flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/resources/KubernetesSharedInformer.java:281]: i mean the loop by "The execution thread will only survive"

Comment (dmvk) [reply flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/resources/KubernetesSharedInformer.java:281]: I think there is a rare case where it won't, basically if you poll between `handling.getAndIncrement() > 0` and 
`queue.put(handlerConsumer);`. Also if `queue.put(handlerConsumer)` would get interrupted in case `present` is larger than `0` it may result in infinite busy waiting loop.

Comment (yittg) [reply flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/resources/KubernetesSharedInformer.java:281]: i think execute callbacks one by one and a new Runnable for each one needs another thread to maintain the order also.

Comment (dmvk) [reply flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/resources/KubernetesSharedInformer.java]: > i think execute callbacks one by one and a new Runnable for each one needs another thread to maintain the order also.

BlockingQueue where you store incoming callbacks from a single thread (this maintains order) + lock (doesn't need to be fair, ordering is maintained by the queue, we just need an exclusive access), inside which you poll callback from the queue and execute it, should be enough.

---
Problem with the previous implementation was, that poll from the executor service queue, was not guarded by the lock.

Comment (yittg) [reply flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/resources/KubernetesSharedInformer.java]: Yeah, it's also ok. i think it still some guard to make the single thread enter and exit.

Comment (yittg) [reply flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/resources/KubernetesSharedInformer.java]: So, @dmvk would you mind i keep the current implementation?

**Comment by wangyang0918** [comment]
@yittg @dmvk I have to admit that using a dedicated thread for each watch will make the code easier to read and maintain. IIRC, each job will need three threads for watching the HA related ConfigMaps. It is reasonable and does not take too much overhead.

File: flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/resources/KubernetesSharedInformer.java
Comment (dmvk) [reply flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/resources/KubernetesSharedInformer.java]: Personally i find it harder to think and reason about as incrementing number of elements in the queue and actually putting and element into the queue is not atomic.

I'd personally prefer something like:

```
Preconditions.checkState(
      callbackQueue.add(handlerConsumer), "Unable to put callback into a queue.");
executorService.submit(
      () -> {
          callbackLock.lock();
          try {
              Objects.requireNonNull(
                              callbackQueue.poll(), "Callback queue is empty.")
                      .accept(handler);
          } finally {
              callbackLock.unlock();
          }
      });
```

But that being said, I don't think I have a strong objection against your solution. WDYT @wangyang0918 ?

**Review by dmvk — APPROVED** [review]
LGTM, this should fix the race condition introduced by executor service internal queue 👍 I've personally prefer different approach for synchronization as outlined above, but I'll leave it up to you and @wangyang0918 to decide.

Good job finding this! 👍

**Comment by dmvk** [comment]
> I have to admit that using a dedicated thread for each watch will make the code easier to read and maintain. IIRC, each job will need three threads for watching the HA related ConfigMaps. It is reasonable and does not take too much overhead.

Which solution are you referring to?

**Comment by wangyang0918** [comment]
> > I have to admit that using a dedicated thread for each watch will make the code easier to read and maintain. IIRC, each job will need three threads for watching the HA related ConfigMaps. It is reasonable and does not take too much overhead.
> 
> Which solution are you referring to?

Given that our original purpose of FLINK-22054 is to save the network connections to K8s APIServer.

What I mean is passing a dedicated single thread when call the `KubernetesSharedInformer#watch`. The current implementation could fix the race condition, however it really makes the code not very easy to read.

**Comment by yittg** [comment]
Thanks @wangyang0918 @dmvk ;
So here are three options if i understand correctly?

1. Revert the previous merged change;
2. Fix with simple lock solution provided by @dmvk; which i think will lead to many thread if events swarm but it is expected not happen normally. And another confuse point is the submit runnable will handle event that may not be corresponding with the current event.
3. Fix with the current PR , i have to admit that it need some thinking to read it.

So would you vote it? WDYT @tillrohrmann

**Comment by wangyang0918** [comment]
> Thanks @wangyang0918 @dmvk ;
> So here are three options if i understand correctly?
> 
> 1. Revert the previous merged change;
> 2. Fix with simple lock solution provided by @dmvk; which i think will lead to many thread if events swarm but it is expected not happen normally. And another confuse point is the submit runnable will handle event that may not be corresponding with the current event.
> 3. Fix with the current PR , i have to admit that it need some thinking to read it.
> 
> So would you vote it? WDYT @tillrohrmann

I would prefer the option 4.

Passing a dedicated single thread, instead of the current shared cached thread pool, when call the `KubernetesSharedInformer#watch`.

File: flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/resources/KubernetesSharedInformer.java
Comment (wangyang0918) [inline flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/resources/KubernetesSharedInformer.java:279]: We need to use the `execute` instead of `submit` here. Right? Because it can swallow exceptions.

**Review by wangyang0918 — CHANGES_REQUESTED** [review]

File: flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/resources/KubernetesSharedInformer.java
Comment (yittg) [reply flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/resources/KubernetesSharedInformer.java]: @dmvk After more thinking, I now prefer your solution. If the caller uses a shared single-threaded executor to call watch, the current implementation may cause the thread to always be occupied by a single watcher. Your solution can avoid this scenario.

**Comment by yittg** [comment]
> I would prefer the option 4.
> 
> Passing a dedicated single thread, instead of the current shared cached thread pool, when call the `KubernetesSharedInformer#watch`.

@wangyang0918 Did you mean change the executor service initialized in `KubernetesHaServices`? @dmvk and i discussed before about leaving the order guarantee on the caller side. But we think it should be guaranteed no matter what executor the caller uses.

File: flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/resources/KubernetesSharedInformer.java
Comment (wangyang0918) [inline flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/resources/KubernetesSharedInformer.java:275]:
```suggestion
                            Preconditions.checkNotNull(callbackQueue.poll(), "Callback queue is empty")
```

**Review by wangyang0918 — COMMENTED** [review]
Given that we do not expect to have too many events, I am not against with current implementation, which will let the `WatchCallback` to guarantee the events are processed orderly.

Could you please verify that how many threads(in the cached thread pool for watching) we will have when running a HA enabled Flink application?

**Comment by yittg** [comment]
> Could you please verify that how many threads(in the cached thread pool for watching) we will have when running a HA enabled Flink application?

@wangyang0918 I'm not quit sure where is the proper place to do the verification. Is 'test_kubernetes_application_ha.sh' the right place? Do you have some good idea?

**Comment by wangyang0918** [comment]
@yittg Maybe I do not make myself clear. What I means is to do the manual verification via `jstack` in the JobManager pod to check the thread num of cached thread pool `config-map-watch-handler`. We need to build a customized image which is based on `openjdk:8-jdk`, not the `openjdk:8-jre`.

I will do it.

**Comment by wangyang0918** [comment]
I have verified in a HA-enabled session cluster with 20 running jobs, the thread num of `config-map-watch-handler` is always smaller than 6, which means we could save the threads and network connections via using a shared cached thread pool.

I have also run the tests in `KubernetesSharedInformerITCase` more than 100 times without failure.

I will merge this PR once the azure pipeline gives green paas.

**Review by wangyang0918 — APPROVED** [review]

**Comment by wangyang0918** [comment]
@yittg The last one comment. We need to factor out the `executorService.submit` -> `executorService.execute` related changes into a separate `hotfix` commit. Once this is done, I will merge this PR.

**Comment by yittg** [comment]
@wangyang0918 I've split this PR into two commit.

**Comment by wangyang0918** [comment]
> @wangyang0918 I've split this PR into two commit.

Thanks. 

nit: Usually, we have the `hotfix` commit in the first place.

**Comment by yittg** [comment]
haha, we pushed the update almost at the same time :), PTAL

**Comment by wangyang0918** [comment]
It looks really good now. I will get it merged tomorrow.

**Comment by yittg** [comment]
@flinkbot run azure

**Comment by yittg** [comment]
Thanks @wangyang0918 , @dmvk

**Pull request closed without merging**
